### ISSUE TITLE
monorepo dependency missing from sourceFiles after using `createProgram`

### ISSUE THREAD

**Issue by thepassle** [microsoft/TypeScript#46360]
# Bug Report

### 🔎 Search Terms

monorepo dependency createProgram

### 🕗 Version & Regression Information
4.4.4

### 💻 Code

I have a simple monorepo with 2 packages:
```
monorepo/
├─ packages/
│  ├─ foo/
│  │  ├─ index.js
│  ├─ bar/
│  │  ├─ index.js
```

where `foo/index.js`:
```js
import { bar } from '@monorepo/bar';

const foo = 'foo' + bar;

console.log(foo);
```

and where `bar/index.js`:
```js
export const bar = 'bar';
```

I'd like to write a script to do some static analysis on `foo/index.js` and all its dependencies:
```js
import ts from 'typescript';

const program = ts.createProgram(['./index.js'], 
  { 
    allowJs: true, 
    moduleResolution: 2
  }
);

program.getSourceFiles()
  .map(({path}) => path)
  .forEach((p) =>{
    console.log(p);
  });
```

Which outputs:
```
/users/blank/monorepo/node_modules/typescript/lib/lib.d.ts
/users/blank/monorepo/node_modules/typescript/lib/lib.es5.d.ts
/users/blank/monorepo/node_modules/typescript/lib/lib.dom.d.ts
/users/blank/monorepo/node_modules/typescript/lib/lib.webworker.importscripts.d.ts
/users/blank/monorepo/node_modules/typescript/lib/lib.scripthost.d.ts
/users/blank/monorepo/packages/foo/index.js
```

Even though `foo/index.js` has a hard dependency on `@monorepo/bar`, `/users/blank/monorepo/package/bar/index.js` is missing from the output of `getSourceFiles`.

If I add any other 'regular' dependency (read: not a monorepo dependency), it does correctly get output from `getSourceFiles`:

`foo/index.js`:
```js
import { bar } from '@monorepo/bar';
import { parse } from 'comment-parser';

parse(`foo`);
const foo = 'foo' + bar;

console.log(foo);
```

output of `getSourceFiles()`:
```
/users/au87xu/monorepo/node_modules/typescript/lib/lib.d.ts
/users/au87xu/monorepo/node_modules/typescript/lib/lib.es5.d.ts
/users/au87xu/monorepo/node_modules/typescript/lib/lib.dom.d.ts
/users/au87xu/monorepo/node_modules/typescript/lib/lib.webworker.importscripts.d.ts
/users/au87xu/monorepo/node_modules/typescript/lib/lib.scripthost.d.ts
/users/au87xu/monorepo/node_modules/comment-parser/lib/primitives.d.ts
/users/au87xu/monorepo/node_modules/comment-parser/lib/parser/tokenizers/index.d.ts
/users/au87xu/monorepo/node_modules/comment-parser/lib/parser/index.d.ts
/users/au87xu/monorepo/node_modules/comment-parser/lib/parser/tokenizers/description.d.ts
/users/au87xu/monorepo/node_modules/comment-parser/lib/parser/tokenizers/name.d.ts
/users/au87xu/monorepo/node_modules/comment-parser/lib/parser/tokenizers/tag.d.ts
/users/au87xu/monorepo/node_modules/comment-parser/lib/parser/tokenizers/type.d.ts
/users/au87xu/monorepo/node_modules/comment-parser/lib/transforms/index.d.ts
/users/au87xu/monorepo/node_modules/comment-parser/lib/transforms/align.d.ts
/users/au87xu/monorepo/node_modules/comment-parser/lib/transforms/indent.d.ts
/users/au87xu/monorepo/node_modules/comment-parser/lib/transforms/crlf.d.ts
/users/au87xu/monorepo/node_modules/comment-parser/lib/util.d.ts
/users/au87xu/monorepo/node_modules/comment-parser/lib/stringifier/inspect.d.ts
/users/au87xu/monorepo/node_modules/comment-parser/lib/stringifier/index.d.ts
/users/au87xu/monorepo/node_modules/comment-parser/lib/index.d.ts
/users/au87xu/monorepo/packages/foo/index.js
```

### 🙁 Actual behavior

`getSourceFiles()` does not include `/users/blank/monorepo/package/bar/index.js`

### 🙂 Expected behavior

`getSourceFiles()` does includes `/users/blank/monorepo/package/bar/index.js`

### Screenshots for clarification:
![image (2)](https://user-images.githubusercontent.com/17054057/137315812-25af9b1c-17f6-4743-8b98-d56438054e5e.png)

![New Project](https://user-images.githubusercontent.com/17054057/137316252-4b947d49-45ca-4752-afe3-a9379667a552.png)

**Comment by MartinJohns** [comment]
By default TypeScript will only lookup relative or in node_modules, and you didn't adjust this logic. GetSourceFiles won't list files it can't find.

**Comment by thepassle** [comment]
> To my knowledge the compiler API is very little documented.

Yeah, additionally I find what little documentation there is very hard to parse and make sense of.

Still though, `@monorepo/bar` is a true dependency of `foo/index.js` and part of its program, so I'd expect it to be listed by `getSourceFiles` (similar to how 'regular' dependencies do get listed), so I still consider this to be a bug.

**Comment by andrewbranch** [comment]
The compiler API can do everything the compiler can do if you use it right, but it’s not clear to me that you have the correct setup even for just a normal compilation with `tsc`. I would suggest creating a tsconfig.json file and using `tsc --listFilesOnly` first so you can make sure tsc is even seeing the right thing, then work on making your API program see the same thing that tsc is seeing.

**Comment by thepassle** [comment]
The fact that its not in `process.cwd() + 'node_modules/'` doesn't mean its not a dependency

**Comment by andrewbranch** [comment]
Did you try with `--maxNodeModulesJsDepth 1` or `2`?
